### PR TITLE
Make sqlite database upgradabale and crypto-agile

### LINKED ISSUE #109: Change decryption key storage format
Currently the decryption key stored in the database is assumed to have a specific format and there is no algorithmic agility.
Furthermore individual encrypted entries do not have any indicator about what key generation was used to encrypt them so it is not easy to perform rolling key changes.

We should change the stored wrapped key to have more structure.
An outer layer should indicate what algorithm(or mechanism version) needs to used to unwrap the stored key.

An inner layer should add known data to the payload to ensure that if a mechanism that does not perform authentication of the encryption is used we still have a way to verify we obtained back a proper key and not just random garbage. And it should add a key generation id to be used for decryption.
If the key generations is formed as a UUID then this can be stored on the key object (as CKA_LABEL) and used as the inner data to confirm proper decryption.

For encrypted entries we should add a key generation id to their outer layer so that the correct key can be used.

Finally a way to store multiple key generations should be added, we have two options:
- Change the inner layer of the wrapped keys to hold multiple generations by treating the payload as an array.
- Allow to store multiple entries per user with historic keys.

The latter would allow to more easily purge older key generations.

### PR DESCRIPTION
**Pull request by simo5**
Add metadata to the sqlite database that allows for easier crypto agility as well as allow future upgrades. Move all non-object information to a metadata table.

Fixes #109

### CODE REVIEW

File: src/storage/sqlite.rs [hunk 1]
```diff
@@ -368,6 +461,160 @@ impl StorageRaw for SqliteStorage {
...
+            Some(&flags32.to_le_bytes() as &[u8]),
+        )?;
+
+        tx.commit().map_err(bad_storage)
+    }
+
+    fn fetch_user(&self, uid: &str) -> Result<StorageAuthInfo> {
+        let conn = self.conn.lock()?;
+        let mut stmt = conn
+            .prepare("SELECT name, data from meta WHERE ((name=? OR name=? OR name=?) AND value=?)")?;
```
Comment (Jakuje) [inline src/storage/sqlite.rs:549]: `name in (?, ?, ?)` SQL would be more appropriate here:

https://www.techonthenet.com/sqlite/in.php

Not sure if the rustqlite have some even better shorthands for the construction of list to make it even simpler.

File: doc/storage-encryption.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+decrypt the storage Encryption key. The Encryption key is versioned and is used
+to encrypt each secret attribute individually by using a key derivation step to
+obtain an object specific key. The Key Derivation algorithms and paramters are
+felxibly defined by wrapping the keys and encrypted values using PKCS#5 like
+ASN.1 structures.
+
+This OID has been defined to identify the HKDF-Expand Key Derivation function:
+1.3.6.1.4.1.2312.20.1.1
+
+KKBPS - Kryoptic Key Based Protection Scheme v1
```
Comment (Jakuje) [inline doc/storage-encryption.md:29]: Every time I read the shortcut `kbps` in lower case I read it as unit of transfer speed kB/s, which is a bit confusing. But I do not have better naming suggestion ...

File: src/storage/format.rs [hunk 3]
```diff
@@ -231,45 +158,24 @@ impl StdStorageFormat {
...
-        let auth_obj = self.aci.make_auth_object(
-            facilities,
-            &get_pin_uid(CKU_SO)?,
-            checked_pin(&[]),
-        )?;
-        self.store.store_obj(auth_obj)
+        let data =
+            self.aci
+                .key_to_user_data(facilities, SO_ID, checked_pin(&[]))?;
+        /* 1 to mark this is the default value */
```
[the remark this answers is not in the document]
Comment (simo5) [reply src/storage/format.rs:164]: yup

File: src/storage/sqlite.rs (hunk 1 above)
Comment (simo5) [reply src/storage/sqlite.rs:549]: no only has parameters, it is a very plain shim on top if libsqlite3

File: doc/storage-encryption.md (hunk 2 above)
Comment (simo5) [reply doc/storage-encryption.md:29]: I know, there isn't much we can do

File: src/storage/aci.rs [hunk 4]
```diff
@@ -50,19 +451,15 @@ impl StorageACI {
...
-        let mut template = CkAttrs::with_capacity(3);
-        template.add_ulong(CKA_CLASS, &class);
-        template.add_ulong(CKA_KEY_TYPE, &keytyp);
-        template.add_ulong(CKA_VALUE_LEN, &keylen);
-        template.add_bool(CKA_EXTRACTABLE, &truebool);
-        let aes = facilities.mechanisms.get(CKM_AES_KEY_GEN)?;
-        self.kek = Some(aes.generate_key(
+        /* Need to use a generic secret here, because the secret key
+         * is used together with HKDF to derive the actual encryption
+         * key and HKDF allows only CKK_HKDf or CKK_GENERIC_SECRET */
```
Comment (Jakuje) [inline src/storage/aci.rs:456]:
```suggestion
         * key and HKDF allows only CKK_HKDF or CKK_GENERIC_SECRET */
```

**Review by Jakuje — APPROVED** [review]
Looks good.
By any chance, did you test this change with pkcs11-provider locally to make sure this works correctly and we wont break the CI?

**Comment by simo5** [comment]
> Looks good. By any chance, did you test this change with pkcs11-provider locally to make sure this works correctly and we wont break the CI?

Yes I always make a run with pkcs11-provider locally to ensure changes actually work in a backwards compatible way.
pkcs11-provider CI more than once caught things.

**Pull request merged**